### PR TITLE
Fixing of grammar and better queries

### PR DESCRIPTION
**Pull request by amaanq**
This PR addresses a couple of issues, both objective and subjective, where the grammar would fail on certain files in the awslabs/smithy repo, and where I found there to be inconsistent/improper grammar structure. I tried to adhere to the IDL as much as possible, as weird as it was as times. 

I also fixed the queries up to more properly reflect nvim-treesitter's standards. I am hoping if this gets merged then this can be added to nvim-treesitter. I am willing to maintain the parsers/grammar.

In addition, I added a test, some examples, and a CI action to run tests.

### CODE REVIEW

**Comment by amaanq** [comment]
Should I not hear from a maintainer for a while, or for whatever reason this wouldn't be merged, I would ideally add my fork to nvim-treesitter then rather than this

File: grammar.js
Comment (keynmol) [inline grammar.js:32]: Why is this `($)` necessary?
I've looked at several other grammars and it's not being used anywhere, so I would suggest reverting it - especially given it makes the diff much larger than it needs to be.

In general, I really like the CI and tests part, but the grammar changes seem to be dramatic and obscured by the `$ => ($)` addition. Could you please revert that so it's clear what you are actually changing?

Here's what I'm talking about: 
image

This became part of a larger diff even though the sections are basically the same.

**Comment by keynmol** [comment]
> Should I not hear from a maintainer for a while

On a related note, @amaanq, I really don't understand this:

1. You created an issue on a repo asking to maintain it, even though it's clear there isn't much activity or backlog - not something that requires extra maintenance
2. You submit a PR and immediately (within 2 minutes of opening the PR) set an ultimatum that either this PR gets merged, or you submit your own fork.

I don't understand what gave you the impression that this repo is unmaintained, given that the PR to nvim-treesitter is handled by another person and is in progress.

More than happy to work with you on this PR and addition to nvim-treesitter, but this ultimatum tone is absolutely unnecessary.

**Comment by amaanq** [comment]
> I don't understand what gave you the impression that this repo is unmaintained, given that the PR to nvim-treesitter is handled by another person and is in progress.

The PR is pretty stale with zero activity from the person who submitted it

> More than happy to work with you on this PR and addition to nvim-treesitter, but this ultimatum tone is absolutely unnecessary.

I agree, I apologize for my tone, in my head your response to my issue seemed like a denial for me to be a maintainer, which I understand as there is not much maintenance needed on this. So again, I do apologize for that.

That being said, either way I would love to get this added into nvim-treesitter :slightly_smiling_face:

File: grammar.js
Comment (amaanq) [reply grammar.js:73]: I find that a bit unnecessarily wordy, especially in the context of parsing output. Sure, for the top level items it makes sense, but imo these are all more definitions rather than statements, and it makes more sense to either suffix it with definition or leave it be (like I have)

File: grammar.js
Comment (amaanq) [reply grammar.js:32]: It stems from the eslint linter/code formatter I added, that is just the spec they use. I'lll revert it now

**Comment by amaanq** [comment]
do you mind if I add a CI and discord badge in the docs? like so:
image

**Comment by amaanq** [comment]
with the latest commit escape sequences can be differentiated with a colorscheme like so

image

File: Cargo.toml
Comment (keynmol) [inline Cargo.toml:7]: While you are here, would you mind please updating the repository URL?
Not that it's used anywhere, but still :)

File: grammar.js
Comment (keynmol) [inline grammar.js:6]: Do we gain anything from extracting `primitives` here?

From what I can see, it's only used in one place (cool trick with `...primitives`, btw, didn't know about it) - and in there it's just the same list.

Comment (amaanq) [reply grammar.js:6]: lots of other grammars use this style, and imo for anyone new to the language wanting to read through the grammar, having primitives at the top is a good start for anyone to gain some understanding about the primitive types

File: Cargo.toml
Comment (amaanq) [reply Cargo.toml:7]: sure sure

File: grammar.js
Comment (keynmol) [reply grammar.js]: I think this goes against the desire you stated in your PR description:

> I tried to adhere to the IDL as much as possible

Which I agree with btw - I don't think the actual wordiness of nodes adds any overhead, but it's much clearer when comparing ABNF with the grammar.

Comment (amaanq) [reply grammar.js]: Yeah, this was one spot I felt very conflicted about. But I agree, maybe this should be taken up with aws then haha. I'll readd the statements in thoes names

File: grammar.js
Comment (keynmol) [reply grammar.js:6]: Gotcha, I still don't see much value in it, but I don't feel strongly about it and am happy to follow the lead from other grammars.

**Comment by amaanq** [comment]
> do you mind if I add a CI and discord badge in the docs? like so: image
> 
> In my grammars I also have a publish action that is run on every tagged commit. See here: https://github.com/amaanq/tree-sitter-kdl/blob/f83f3943568c7e7b4f5e0de1b04d722223bd4d80/.github/workflows/publish.yml
> 
> So if a commit is tagged v0.4.0 or whatever and pushed, this CI action will be run then and only then, all you'd have to do is sync the git tag with the cargo and npm manifest versions.
> 
> This would be a good idea to add but totally your choice

@keynmol Thoughts on this? Seems like this is the last topic to discuss/resolve

**Comment by keynmol** [comment]
> @keynmol Thoughts on this? Seems like this is the last topic to discuss/resolve

Badges - go ahead ✅

With regards to publishing - I never really considered that (I also maintain tree-sitter-scala, which has a lot more users) - do you have any examples of applications/libraries that use NPM packages or crates?

I'd be happy to look into setting up the secrets for publishing, but it will be lower on priority list.

**Comment by amaanq** [comment]
> > @keynmol Thoughts on this? Seems like this is the last topic to discuss/resolve
> 
> Badges - go ahead white_check_mark
> 
> With regards to publishing - I never really considered that (I also maintain tree-sitter-scala, which has a lot more users) - do you have any examples of applications/libraries that use NPM packages or crates?
> 
> I'd be happy to look into setting up the secrets for publishing, but it will be lower on priority list.

Those that want to write their own editor or tool to integrate tree-sitter parsing with tend to use the npm package or rust crate instead of the bare cli; I've had requests for publishing mine before as someone wanted that programmatic usage out of the parser.

**Review by keynmol — APPROVED** [review]
In the current state it looks great - I'm especially impressed with the highlight queries and CI, thank you so much!

For the time being it's okay to check-in all the auto-generated files, but I will setup https://github.com/indoorvivants/tree-sitter-smithy/issues/8 in the near future. 

Happy to merge this once you add the badges.

**Comment by amaanq** [comment]
> In the current state it looks great - I'm especially impressed with the highlight queries and CI, thank you so much!

Thanks, glad I could contribute :) 

For #8, I'm assuming that's so that the large diffs aren't rendered in someone's PR/commits? The .gitattributes file solves exactly that. If you mean in case someone forgets to generate when making changes, then of course I'm all for that, and that'd be a good idea to add to my parsers too...good idea :slightly_smiling_face:

**Comment by keynmol** [comment]
> For #8, I'm assuming that's so that the large diffs aren't rendered in someone's PR/commits? The .gitattributes file solves exactly that. If you mean in case someone forgets to generate when making changes, then of course I'm all for that, and that'd be a good idea to add to my parsers too...good idea 🙂

It's mostly for grammars with lots of traffic - you can read justification here: https://github.com/tree-sitter/tree-sitter-scala/issues/85

Basically it's to

1. Reduce PR conflicts - we had lots of PRs in parallel and they can be automatically resolved by git only on smaller, simpler files (like tests and grammar.js)
2. Test in a pristine (ish) CI environment - no user checked-in C code
3. If user checks in C code, verify that CI produces exactly the same one
4. (additionally) we also verify that the queries in nvim-treesitter aren't broken by the PRs in the grammar repo - to avoid breaking the users

IMO that's just better way to vendor C code generated by parsers, and make reviews easier as well. I shiver uncontrollably when I see diffs like this 
image

and would rather set up a process where nobody has to check that the C code is correct :)

**Comment by amaanq** [comment]
> > For #8, I'm assuming that's so that the large diffs aren't rendered in someone's PR/commits? The .gitattributes file solves exactly that. If you mean in case someone forgets to generate when making changes, then of course I'm all for that, and that'd be a good idea to add to my parsers too...good idea 🙂
> 
> It's mostly for grammars with lots of traffic - you can read justification here: https://github.com/tree-sitter/tree-sitter-scala/issues/85
> 
> Basically it's to
> 
> 1. Reduce PR conflicts - we had lots of PRs in parallel and they can be automatically resolved by git only on smaller, simpler files (like tests and grammar.js)
> 2. Test in a pristine (ish) CI environment - no user checked-in C code
> 3. If user checks in C code, verify that CI produces exactly the same one
> 4. (additionally) we also verify that the queries in nvim-treesitter aren't broken by the PRs in the grammar repo - to avoid breaking the users
> 
> IMO that's just better way to vendor C code generated by parsers, and make reviews easier as well. I shiver uncontrollably when I see diffs like this 
> image
> 
> and would rather set up a process where nobody has to check that the C code is correct :)

That's awesome! But what if someone submitting a PR generated the code themselves without knowing about this, the diff would still be huge then right?

**Comment by keynmol** [comment]
> That's awesome! But what if someone submitting a PR generated the code themselves without knowing about this, the diff would still be huge then right?

Yeah, but then the CI can verify that C code is produced by the version of the grammar in the PR (as I care that they're in sync), and the PR author can just force-push and remove those files (when [asked](https://github.com/tree-sitter/tree-sitter-scala/blob/master/CONTRIBUTING.md#generation)) :)

Being able to work with both the a) PRs that contain generated code and b) PRs without it, is key to flexible contribution process - and in the end you still uphold the guarantee that grammar and parser are fully in sync.

Overall it can involve an extra step if the PR author is used to the way it's done on other grammars, but IMO it's worth it to keep PRs individually revertible and cherry-pickable.

**Comment by keynmol** [comment]
Thanks again for awesome contribution!

**Comment by keynmol** [comment]
> Just to double check, can I add you as a co-maintainer of smithy alongside me for nvim-treesitter?

Yep.

**Pull request merged**
